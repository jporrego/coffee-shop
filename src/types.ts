export interface Product {
  id: number;
  name: string;
  price: number;
  description: string;
  img: string;
}

export type StringTouple = [string, string];
