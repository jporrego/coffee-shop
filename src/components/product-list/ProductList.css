.product-list {
    display: flex;
    justify-content: center;
    align-content: flex-start;
    flex-wrap: wrap;
    width: 100%;
    gap: 2rem;
    padding: 1rem 0rem 3rem 0rem;
    z-index: 0;
}

@media only screen and (min-width: 600px) {
    .product-list {
        justify-content: flex-start;
    }
}