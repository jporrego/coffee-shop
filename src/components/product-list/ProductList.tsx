import React, { useState, useEffect } from "react";
import "./ProductList.css";
import { Product as ProductInterface } from "../../types";
import Product from "../product/Product";
import Spinner from "../../assets/spinner.svg";

interface ProductListProps {
  products: ProductInterface[];
  setProducts: React.Dispatch<React.SetStateAction<ProductInterface[]>>;
  onAddToCart: (id: string) => void;
  openProductModal: (id: string) => void;
}
const ProductList: React.FC<ProductListProps> = ({
  products,
  setProducts,
  onAddToCart,
  openProductModal,
}) => {
  const [loading, setLoading] = useState<boolean>(false);
  const [errorMsg, setErrorMsg] = useState<string>("");

  useEffect(() => {
    getProducts();
  }, []);

  const getProducts = async () => {
    /*
    if (process.env.REACT_APP_API_URL !== undefined) {
      try {
        setLoading(true);
        setErrorMsg(
          "Starting Heroku server. This can take around 10 seconds..."
        );
        const response = await fetch(process.env.REACT_APP_API_URL);
        const data = await response.json();
        setLoading(false);
        setErrorMsg("");
        setProducts(data);
      } catch (error) {
        console.log(error);
        setLoading(false);
        setErrorMsg("Failed to connect to the server.");
      }
    } else {
      setErrorMsg("Failed to connect to the server.");
    }*/
    let tryToFetch = true;
    setTimeout(() => {
      tryToFetch = false;
    }, 20000);
    while (tryToFetch) {
      if (process.env.REACT_APP_API_URL !== undefined) {
        try {
          setLoading(true);
          setErrorMsg(
            "Starting Heroku server. This can take around 10 seconds..."
          );
          const response = await fetch(process.env.REACT_APP_API_URL);
          const data = await response.json();
          setProducts(data);

          setLoading(false);
          setErrorMsg("");
          break;
        } catch (error) {
          console.log(error);
          setLoading(false);
          if (tryToFetch) {
            setErrorMsg(
              "Starting Heroku server. This can take around 10 seconds..."
            );
          } else {
            setErrorMsg("Failed to connect to the server.");
          }
        }
      } else {
        setErrorMsg("Failed to connect to the server.");
        break;
      }
    }
  };

  return (
    <div className="product-list">
      {loading && (
        <div className="spinner">
          <img src={Spinner} alt="React Logo" />
        </div>
      )}
      <div className="error-msg">{errorMsg}</div>
      {products.map((product) => (
        <Product
          key={product._id}
          product={product}
          onAddToCart={onAddToCart}
          openProductModal={openProductModal}
        ></Product>
      ))}
    </div>
  );
};

export default ProductList;
