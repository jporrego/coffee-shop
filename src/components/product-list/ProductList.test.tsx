import React from "react";
import { queryAllByTestId, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import ProductList from "./ProductList";
import { Product as ProductInterface } from "../../types";

interface MockProductProps {
  product: ProductInterface;
  onAddToCart: typeof jest.fn;
  openProductModal: typeof jest.fn;
}

jest.mock(
  "../product/Product",
  () =>
    ({ product, onAddToCart, openProductModal }: MockProductProps) =>
      <div data-testid="product">{product.name}</div>
);

// set ProductList props
const products: ProductInterface[] = [
  {
    id: 0,
    name: "Product",
    price: 10,
    img: "breville.png",
    description: "Product description",
  },
  {
    id: 1,
    name: "Product 2",
    price: 20,
    img: "nesspresso-kups.png",
    description: "Product 2 description",
  },
];

const setProducts = jest.fn();
const onAddToCart = jest.fn();
const openProductModal = jest.fn();

describe("product list", () => {
  it("renders the product list", () => {
    const { container } = render(
      <ProductList
        products={products}
        setProducts={setProducts}
        onAddToCart={onAddToCart}
        openProductModal={openProductModal}
      ></ProductList>
    );

    expect(container).toMatchSnapshot();
  });

  it("renders two products", () => {
    render(
      <ProductList
        products={products}
        setProducts={setProducts}
        onAddToCart={onAddToCart}
        openProductModal={openProductModal}
      ></ProductList>
    );

    expect(screen.queryAllByTestId("product").length).toBe(2);
  });
});
