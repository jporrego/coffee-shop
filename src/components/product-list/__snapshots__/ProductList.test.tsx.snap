// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`product list renders the product list 1`] = `
<div>
  <div
    class="product-list"
  >
    <div
      data-testid="product"
    >
      Product
    </div>
    <div
      data-testid="product"
    >
      Product 2
    </div>
  </div>
</div>
`;
